src/openlcb/BroadcastTimeDefs.cxx

```diff
@@ -0,0 +1,161 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file BroadcastTimeDefs.cxx
+ *
+ * Static definitions for implementations of the OpenLCB Broadcast Time
+ * Protocol.
+ *
+ * @author Stuart W. Baker
+ * @date 30 July 2022
+ */
+
+#include "BroadcastTimeDefs.hxx"
+
+#include <string>
+
+namespace openlcb
+{
+
+extern "C"
+{
+// normally requires _GNU_SOURCE
+char *strptime(const char *__restrict, const char *__restrict,
+               struct tm *__restrict);
+}
+
+//
+// BroadcastTimeDefs::time_to_string()
+//
+std::string BroadcastTimeDefs::time_to_string(int hour, int min)
+{
+    if (hour < 0 || hour > 23)
+    {
+        hour = 0;
+    }
+    if (min < 0 || min > 59)
+    {
+        min = 0;
+    }
+
+    struct tm tm;
+    tm.tm_hour = hour;
+    tm.tm_min = min;
+    char value[6];
+    if (strftime(value, 6, "%R", &tm) != 0)
+    {
+        return value;
+    }
+    else
+    {
+        return "Error";
+    }
+}
+
+//
+// BroadcastTimeDefs::rate_quarters_to_string()
+//
+std::string BroadcastTimeDefs::rate_quarters_to_string(int16_t rate)
+{
+    std::string result;
+    if (rate < 0)
+    {
+        result.push_back('-');
+        rate = -rate;
+    }
+    uint16_t whole = rate >> 2;
+    uint16_t frac = rate & 0x3;
+    if (whole >= 100)
+    {
+        result.push_back('0' + (whole / 100));
+    }
+    if (whole >= 10)
+    {
+        result.push_back('0' + ((whole % 100) / 10));
+    }
+    result.push_back('0' + (whole % 10));
+    switch (frac)
+    {
+        default:
+        case 0:
+            result.append(".00");
+            break;
+        case 1:
+            result.append(".25");
+            break;
+        case 2:
+            result.append(".50");
+            break;
+        case 3:
+            result.append(".75");
+            break;
+    }
+    return result;
+}
+
+//
+// BroadcastTimeDefs::string_to_time()
+//
+bool BroadcastTimeDefs::string_to_time(std::string stime, int *hour, int *min)
+{
+    struct tm tm;
+    if (strptime(stime.c_str(), "%R", &tm) != nullptr)
+    {
+        if (hour)
+        {
+            *hour = tm.tm_hour;
+        }
+        if (min)
+        {
+            *min = tm.tm_min;
+        }
+        return true;
+    }
+    return false;
+}
+
+//
+// BroadcastTimeDefs::string_to_rate_quaraters()
+//
+int16_t BroadcastTimeDefs::string_to_rate_quarters(std::string srate)
```

Comment:
consider adding a unit test that exercised these utility functions, including verifying that they behave as expected around the limit values and errors.

Reply:
https://github.com/bakerstu/openmrn/issues/640